### PR TITLE
Interfaces: Add interface between ROS and sitl_gazebo topics

### PR DESCRIPTION
**Pull request by TSC21**
This pull request adds the `gazebo_ros_interface` plugin to this repository. It is based of from [*rotors_simulator* `gazebo_ros_interface_plugin`](https://github.com/ethz-asl/rotors_simulator/blob/master/rotors_gazebo_plugins/src/gazebo_ros_interface_plugin.cpp) (so, the credits and license were kept), but with a enhanced layer of abstraction that allows the plugin to infer over the existing plugins set to each and every model that is spawned in a world and subscribes to the topics that they publish (ex, if I have 2 Iris spawned, I will get the barometer data from both of them published into different ROS topics).

The main reason for this plugin is to fix up the limitations in Gazebo in terms of the introspection tools: Gazebo plot GUI, for example, is unable to get the data from Gazebo topics were the `protobuf` msgs are custom. The same happens when trying to echo these same topics using `gz topic`. The reason is that these protobuf messages we are using are not self-describable - https://bitbucket.org/osrf/gazebo/issues/2569/support-for-self-describing-messages - the support could be added but would require a great effort, which probably wouldn't be justified since Gazebo Ignition (and Ignition Transport) will probably solve these matters.

But, while we still support the "legacy" Gazebo Transport, the best work around is to make possible to have the ROS introspection tools to ease our debugging process with real-time plotting - example. using `rqt_plot` or PlotJuggler - together with the ground station Mavlink inspection tools (in the autopilot side in this case).

I added the subscription to most of the topics that we can find (and are relevant) to the debug process. We can still add, for example, the subscription to the mag topic or the vision topic. But I guess we can leave that for a further iteration.

@m1kelyons FYI.

### CODE REVIEW

**Comment by TSC21** [comment]
Also, a further PR will be done at upstream PX4 Firmware so to enable building and using this plugin using the `make` SITL targets, without the need to build this inside a `catkin_ws`.

**Comment by TSC21** [comment]
CI issue is unrelated

File: include/gazebo_ros_interface.h [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#include <sensor_msgs/Temperature.h>
+#include <sensor_msgs/Range.h>
+
+#include <common.h>
+
+namespace gazebo {
+
+typedef const boost::shared_ptr<const sensor_msgs::msgs::Pressure> GzBaroMsgPtr;
+typedef const boost::shared_ptr<const sensor_msgs::msgs::Imu> GzImuMsgPtr;
+typedef const boost::shared_ptr<const sensor_msgs::msgs::SITLGps> GzSITLGpsMsgPtr;
```
Comment (julianoes) [inline include/gazebo_ros_interface.h:61]:
```suggestion
typedef const boost::shared_ptr<const sensor_msgs::msgs::SitlGps> GzSitlGpsMsgPtr;
```

File: src/gazebo_ros_interface.cpp [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+        "~/" + models[i]->GetName() + "/camera/link/irlock", irlock_ros_publishers, gz_node_handle_);
+    }
+    // Verify if the Lidar (or model specific) sensor joint exists
+    if (std::find(joint_names.begin(), joint_names.end(), "lidar_joint") != joint_names.end() ||
+        std::find(joint_names.begin(), joint_names.end(), "sf10a_joint") != joint_names.end()) {
+      // Create ROS publisher
+      ros::Publisher lidar_pub = ros_node_handle_->advertise<sensor_msgs::Range>("/gazebo/" + models[i]->GetName() + "/lidar/data", 1);
+      std::vector<ros::Publisher> lidar_ros_publishers = {lidar_pub};
+
+      // handle Diff pressure plugin data
```
Comment (julianoes) [inline src/gazebo_ros_interface.cpp:141]: What is the relation between diff pressure and the laser scanner? I don't understand that.
Or is that comment just outdated?

File: src/gazebo_ros_interface.cpp [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+
+  // Publish to ROS
+  ros_publishers[0].publish(imu_msg);
+}
+
+void GazeboRosInterface::GzSITLGpsMsgCallback(GzSITLGpsMsgPtr& gz_sitl_gps_msg, std::vector<ros::Publisher> ros_publishers) {
+  auto fix_msg = boost::make_shared<sensor_msgs::NavSatFix>();
+
+  fix_msg->header.frame_id = "map";
+  fix_msg->header.stamp = ros::Time(gz_sitl_gps_msg->time_usec() * 1000UL / 1000000000UL, gz_sitl_gps_msg->time_usec() * 1000UL % 1000000000UL);
```
Comment (julianoes) [inline src/gazebo_ros_interface.cpp:284]: What about a helper function?
```
ros::Time rosTimeFromUsec(uint64_t usec) {
    return ros::Time(usec() * 1000UL / 1000000000UL, usec() * 1000UL % 1000000000UL);
}
```

File: src/gazebo_ros_interface.cpp (hunk 2 above)
Comment (TSC21) [reply src/gazebo_ros_interface.cpp:141]: ah sorry. outdated

File: src/gazebo_ros_interface.cpp (hunk 3 above)
Comment (TSC21) [reply src/gazebo_ros_interface.cpp:284]: fancy. will use it

**Comment by Jaeyoung-Lim** [comment]
@TSC21 Do you have any thoughts on how we should move this forward?

We have catkin build tests as part of our CI testing, so we should be at a better place to add this

**Comment by TSC21** [comment]
> @TSC21 Do you have any thoughts on how we should move this forward?
> 
> We have catkin build tests as part of our CI testing, so we should be at a better place to add this

I do I just need some time to reboot it and rebase it. Should be trivial. Let me put this on my ToDo for next week

**Comment by LorenzMeier** [comment]
I'm closing this as stale. Please open a new PR if this is still relevant.

**Pull request closed without merging**